chain/epoch-manager/src/shard_tracker.rs

```diff
@@ -340,6 +367,33 @@ impl ShardTracker {
     }
 }
 
+/// Checks whether `shard_id` is a descendant of any of the `tracked_shards`.
+/// Assumes that `shard_id` exists in the shard layout of `epoch_id`.
+fn check_if_descendant_of_tracked_shard(
+    mut shard_id: ShardId,
+    tracked_shards: &Vec<ShardId>,
+    epoch_id: &EpochId,
+    epoch_manager: &Arc<dyn EpochManagerAdapter>,
+) -> Result<bool, EpochError> {
+    if tracked_shards.contains(&shard_id) {
+        // `shard_id` is one of the `tracked_shards` (a shard is a descendant of itself)
+        return Ok(true);
+    }
+    let mut protocol_version = epoch_manager.get_epoch_protocol_version(epoch_id)?;
+    while protocol_version > PROD_GENESIS_PROTOCOL_VERSION {
+        let shard_layout = epoch_manager.get_shard_layout_from_protocol_version(protocol_version);
+        protocol_version -= 1;
+        let Some(parent_shard_id) = shard_layout.try_get_parent_shard_id(shard_id)? else {
+            continue;
+        };
+        if tracked_shards.contains(&parent_shard_id) {
+            return Ok(true);
+        }
+        shard_id = parent_shard_id;
+    }
```

Comment:
This works well for the new shard layouts with unique shard ids. Does it work for the old ones though? In the old shard layouts you can have the same shard id repeated between layouts but referring to different shards e.g. splitting 3-> 3, 4

Reply:
That would track all descendants of each occurence of the given shardId in the resharding history. For `tracked_shards: [3]` and split `3 -> 3, 4` we would track both 3 and 4.
I think it makes sense to track the latest occurence of the shard ID only, i.e. track shard 3 only in the example above.
Changed the behavior.